cfgov/jinja2/v1/_includes/organisms/mega-menu.html

```diff
@@ -255,3 +255,9 @@ <h2 class="{{ base_class ~ '_heading' }}">
 </nav>
 
 {% endmacro %}
+
+
+{# Render the menu if this entire template is rendered with menu_items. #}
```

Comment:
I don't quite understand this. Does this if statement prevent the menu template from rendering in the case where there is no mega menu created in Wagtail? 

I tried deleting the entire (Spanish) menu in wagtail to see this case, and there was an error instead.

Reply:
Good feedback; I realize this comment might not be that clear. Let me try to explain what I mean by this, and maybe together we can come up with better wording.

In the existing mega menu implementation, the template [mega-menu.html](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/mega-menu.html) (and its two proposed variations, [mega-menu-var-1.html](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/mega-menu-var-1.html) and [mega-menu-var-2.html](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/mega-menu-var-2.html)) can't be rendered by Django on its own, that is, using the standard process of loading the template and then calling some kind of `template.render(context)` on it. This is because, in master, the template only contains Jinja macros; it doesn't contain any content at the top level. Currently the top-level entry point is [this](https://github.com/cfpb/cfgov-refresh/blob/3a4a2b9264b2c386b0a2aa0b7ae06138f1f8296f/cfgov/jinja2/v1/_includes/organisms/mega-menu.html#L232) `mega_menu()` macro, which has to be explicitly invoked.

Currently we do that in the header template [header.html](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/header.html). First we use Jinja's [`import`](https://jinja.palletsprojects.com/en/2.11.x/templates/#import) to import that `mega_menu` macro from the template file [here](https://github.com/cfpb/cfgov-refresh/blob/3a4a2b9264b2c386b0a2aa0b7ae06138f1f8296f/cfgov/jinja2/v1/_includes/organisms/header.html#L106), and then we invoke that macro [here](https://github.com/cfpb/cfgov-refresh/blob/3a4a2b9264b2c386b0a2aa0b7ae06138f1f8296f/cfgov/jinja2/v1/_includes/organisms/header.html#L114), passing it the data to render.

What I'd like to do instead, as part of this PR, is be able to "just" render the mega menu template itself, from Python, without having to invoke a specific macro in it. In this PR I expose a new `mega_menu` function to Jinja that renders the whole menu, and invoke it [here](https://github.com/cfpb/cfgov-refresh/blob/1dd2688815ee69f6565ae0dcf325946352af9091/cfgov/jinja2/v1/_includes/organisms/header.html#L94) (this line is now calling [this](https://github.com/cfpb/cfgov-refresh/blob/1dd2688815ee69f6565ae0dcf325946352af9091/cfgov/mega_menu/jinja2tags.py#L51) Jinja function, not a macro from a template). And I want my function to be able to "just" call `template.render(context)` to render the menu, on this line, [here](https://github.com/cfpb/cfgov-refresh/blob/1dd2688815ee69f6565ae0dcf325946352af9091/cfgov/mega_menu/jinja2tags.py#L43). But in order to do that, I have to change the template as I've done here, with this logic:

```jinja
{# Render the menu if this entire template is rendered with menu_items. #}
{% if menu_items %}
    {{ mega_menu( menu_items, language | default( null ) ) }}
{% endif %}
```

What this does is, if the template is `render`ed with a context containing a `menu_items` variable, it will call that top-level macro, to invoke it.

Writing all this out, I fear it seems very complex, but the main idea was to simplify the point where the HTML frontend calls the backend, and let the single new `mega_menu` call in Jinja do all the work. Does that make any sense?

Reply:
That makes some sense. Are you planning on simplifying the templates to remove the `if` conditions after we migrate to this new mega menu back-end? (I'm assuming there will be a lot of cleanup involved after the switch)

Reply:
You're right that there will be a lot of cleanup, but honestly I hadn't thought that much about the template. One reasonable way to do it might be to remove the `if` as you suggest and then add some documentation to the top of the template about how to render it -- that you need to provide `menu_items` and, more importantly, what structure those items need to be. That structure can also hopefully be simplified once we move to the new backend.